kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt

```diff
@@ -17,14 +17,43 @@ package com.squareup.workflow
 
 /**
  * A [Workflow] that has no state.
+ *
+ * @see StatefulWorkflow
  */
-typealias StatelessWorkflow<InputT, OutputT, RenderingT> =
-    Workflow<InputT, Unit, OutputT, RenderingT>
+interface StatelessWorkflow<InputT : Any, OutputT : Any, RenderingT : Any> :
```

Comment:
I don't want to increase the scope of this PR, more a discussion of possible follow up work: I still find myself wondering if the code would be simpler (both for us to maintain and for others to consume) if `StatelessWorkflow` and `StatefulWorkflow` were abstract classes.

Reply:
```kotlin
sealed class Workflow<I, O, R> {
  internal abstract fun asStatefulWorkflow(): StatefulWorkflow<I, *, O, R>
}

abstract class StatefulWorkflow<I, S, O, R> : Workflow<I, O, R>() {
  final override internal fun asStatefulWorkflow() = this

  abstract fun initialState(…
}

abstract class StatelessWorkflow<I, O, R> : Workflow<I, O, R>() {
  final override internal fun asStatefulWorkflow() = object : …
  
  abstract fun compose(I, …
}
```

Reply:
So using abstract classes (without a sealed parent) gives us the ability to make the `asStatefulWorkflow` override final in the 2 children, which would be nice.

Making them _sealed_ lets us get rid of that function entirely. We could have a separate `WorkflowNode` just for stateless workflows, which would be much simpler than the stateful one and could use less memory.

Those are the only two advantages of using abstract classes over interfaces that I can think of. Did you have anything else in mind @rjrjr?

Reply:
Trying to think if there are any disadvantages to using abstract classes…

The only thing is that it limits multiple inheritance, but I'm not sure that matters. In fact, it's probably even a nice benefit since other classes shouldn't be implementing workflow types anyway.

Reply:
Nvm about sealed classes - `Workflow` can't be a class because it would prevent consumers from using the public-interface-private-impl pattern (can't extend both `AuthWorkflow` and `StatefulWorkflow` if they're both classes).

Reply:
So excited!

Reply:
This PR now also makes `StatelessWorkflow` and `StatefulWorkflow` abstract classes!